specifyweb/frontend/js_src/lib/components/Statistics/index.tsx

```diff
@@ -0,0 +1,740 @@
+import { dynamicCategories } from './definitions';
+import {
+  statsToTsv,
+  useBackendApi,
+  setAbsentCategoriesToFetch,
+  useDefaultLayout,
+  useDefaultStatsToAdd,
+  useStatsSpec,
+  useDynamicCategorySetter,
+} from './hooks';
+import { StatsPageEditing } from './StatsPageEditing';
+import type { CustomStat, DefaultStat, StatLayout } from './types';
+import { collectionPreferences } from '../Preferences/collectionPreferences';
+
+export function StatsPage(): JSX.Element | null {
+  useMenuItem('statistics');
+  const [collectionLayout, setCollectionLayout] = collectionPreferences.use(
+    'statistics',
+    'appearance',
+    'layout'
+  );
+
+  const [personalLayout, setPersonalLayout] = userPreferences.use(
+    'statistics',
+    'appearance',
+    'layout'
+  );
+
+  const [defaultLayout, setDefaultLayout] = collectionPreferences.use(
+    'statistics',
+    'appearance',
+    'defaultLayout'
+  );
+
+  const layout = {
+    [statsText.shared()]: collectionLayout,
+    [statsText.personal()]: personalLayout,
+  };
+
+  const [state, setState] = React.useState<
+    | State<
+        'AddingState',
+        {
+          readonly pageIndex: number;
+          readonly categoryIndex: number;
+        }
+      >
+    | State<
+        'DeletingCategoryState',
+        { readonly categoryContainsCustom: boolean }
+      >
+    | State<
+        'PageRenameState',
+        {
+          readonly pageIndex: number | undefined;
+          readonly isCollection: boolean;
+        }
+      >
+    | State<'DefaultState'>
+    | State<'EditingState'>
+  >({ type: 'DefaultState' });
+
+  const isAddingItem = state.type === 'AddingState';
+  const isEditing =
+    state.type === 'EditingState' ||
+    isAddingItem ||
+    state.type === 'PageRenameState';
+  const isEditingCollection = React.useRef(false);
+  const canEditIndex = (isCollection: boolean): boolean =>
+    isCollection
+      ? hasPermission('/preferences/statistics', 'edit_protected') &&
+        isEditingCollection.current
+      : !isEditingCollection.current;
+
+  const [activePage, setActivePage] = React.useState<{
+    readonly isCollection: boolean;
+    readonly pageIndex: number;
+  }>({
+    isCollection: true,
+    pageIndex: 0,
+  });
+
+  const setLayout = activePage.isCollection
+    ? setCollectionLayout
+    : setPersonalLayout;
+  const sourceLayout = activePage.isCollection
+    ? collectionLayout
+    : personalLayout;
+
+  const allCategories = React.useMemo(() => dynamicCategories, []);
+  const [categoriesToFetch, setCategoriesToFetch] = React.useState<RA<string>>(
+    []
+  );
+
+  setAbsentCategoriesToFetch(
+    sourceLayout,
+    categoriesToFetch,
+    setCategoriesToFetch
+  );
+  const backEndResponse = useBackendApi(categoriesToFetch);
+  const defaultBackEndResponse = useBackendApi(allCategories);
+  const statsSpec = useStatsSpec();
+  const defaultLayoutSpec = useDefaultLayout(statsSpec);
+
+  /*
+   * Initial Load For Collection and Personal Pages
```

Comment:
Limit line length to 80 characters.

You can set a visual guide in your IDE for helping:

<img width="1162" alt="image" src="https://user-images.githubusercontent.com/40512816/217631593-a183fdb6-81e9-4abd-8906-16dfc2296928.png">